packages/wrangler/src/pages/dev.ts

```diff
@@ -476,15 +521,80 @@ export const Handler = async (args: PagesDevArguments) => {
 					routesModule,
 					defineNavigatorUserAgent,
 				});
+
+				for (const dep in bundle.dependencies) {
+					const resolvedDep = resolve(functionsDirectory, dep);
+					/*
+					 * EXCLUDE:
+					 *   - the "/functions" directory because we're already watching it
+					 *   - everything in "./.wrangler", as it's mostly cache and
+					 *     temporary files
+					 *   - any bundle dependencies we are already watching
+					 *   - anything outside of the current working directory, since we
+					 *     are expecting `wrangler pages dev` to be run from the Pages
+					 *     project root folder
+					 */
+					if (
+						!resolvedDep.match(/\/functions\/./) &&
+						!resolvedDep.match(/\.wrangler\/./) &&
+						resolvedDep.match(resolve(process.cwd()))
+					) {
+						crrBundleDependencies.add(resolvedDep);
+
+						if (!watchedBundleDependencies.has(resolvedDep)) {
+							watchedBundleDependencies.add(resolvedDep);
+							watcher.add(resolvedDep);
+						}
+					}
+				}
+
+				// handle non-JS module dependencies, such as wasm/html/binary imports
+				for (const module of bundle.modules) {
+					if (module.filePath) {
+						const resolvedDep = resolve(functionsDirectory, module.filePath);
+						crrBundleDependencies.add(resolvedDep);
+
+						if (!watchedBundleDependencies.has(resolvedDep)) {
+							watchedBundleDependencies.add(resolvedDep);
+							watcher.add(resolvedDep);
+						}
+					}
+				}
+
+				/*
+				 *`bundle.dependencies` and `bundle.modules` will always contain the
+				 * latest dependency list of the current bundle. If we are currently
+				 * watching any dependency files not in that list, we should remove
+				 * them, as they are no longer relevant to the compiled Functions.
+				 */
+				watchedBundleDependencies.forEach(async (path) => {
```

Comment:
Does `watcher.getWatched()` work for this use case? I wonder if it would allow you to remove the `watchedBundleDependencies` `Set`

Reply:
FWICT`watcher.getWatched()` seem to be rather reliable (https://github.com/paulmillr/chokidar/issues?q=is%3Aissue+getwatched+is%3Aclosed), which is why I avoided it. But we can do something like `watcher.unwatch([file1, file2, ..., file_n])`, if that's what you meant.

Reply:
Ah, I didn't realise it was so unreliable—in that case your `Set` solution is great!